# Import Libaries

In [137]:
# Import Libaries
import pandas as pd
import numpy as np

# Data Cleaning for Resale Flat data with longitude and latitude information

- The resale flat price are retrieved from the from the following website from 1990 onwards to July 2023 [[1]](https://beta.data.gov.sg/datasets/189/resources/d_8b84c4ee58e3cfc0ece0d773c8ca6abc/view)

- The dataset has been combined with longitude and latitude data extracted from onemap [[2]](https://www.onemap.gov.sg/apidocs/apidocs)

## Import Dataset

In [138]:
# Import in Dataset
resale_long_lat_fmt_only = pd.read_csv('../assets/resale_flat_gen/resale_long_lat_fmt_only.csv')


C:\Users\user\AppData\Local\Temp\ipykernel_2796\2893526200.py:2: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  resale_long_lat_fmt_only = pd.read_csv('../assets/resale_flat_gen/resale_long_lat_fmt_only.csv')


In [139]:
#
resale_long_lat_fmt_only.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 903393 entries, 0 to 903392
Data columns (total 14 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   month                903393 non-null  object 
 1   town                 903393 non-null  object 
 2   flat_type            903393 non-null  object 
 3   block                903393 non-null  object 
 4   street_name          903393 non-null  object 
 5   storey_range         903393 non-null  object 
 6   floor_area_sqm       903393 non-null  float64
 7   flat_model           903393 non-null  object 
 8   lease_commence_date  903393 non-null  int64  
 9   resale_price         903393 non-null  float64
 10  remaining_lease      194343 non-null  object 
 11  address              903393 non-null  object 
 12  longitude            903393 non-null  float64
 13  latitude             903393 non-null  float64
dtypes: float64(4), int64(1), object(9)
memory usage: 96.5+ MB


## Check if there are null values in the dataset

In [140]:
# Check if there are missing values for dataset
resale_long_lat_fmt_only.isnull().sum()

month                       0
town                        0
flat_type                   0
block                       0
street_name                 0
storey_range                0
floor_area_sqm              0
flat_model                  0
lease_commence_date         0
resale_price                0
remaining_lease        709050
address                     0
longitude                   0
latitude                    0
dtype: int64

Missing values is seen for the column "remaining_lease".

In [141]:
# View the top 5 rows 
resale_long_lat_fmt_only.head(5)

month        town flat_type block       street_name storey_range  \
0  1990-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
1  1990-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     04 TO 06   
2  1990-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12   
3  1990-01  ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     07 TO 09   
4  1990-01  ANG MO KIO    3 ROOM   216  ANG MO KIO AVE 1     04 TO 06   

   floor_area_sqm      flat_model  lease_commence_date  resale_price  \
0            31.0        IMPROVED                 1977        9000.0   
1            31.0        IMPROVED                 1977        6000.0   
2            31.0        IMPROVED                 1977        8000.0   
3            31.0        IMPROVED                 1977        6000.0   
4            73.0  NEW GENERATION                 1976       47200.0   

  remaining_lease               address   longitude  latitude  
0             NaN  309 ANG MO KIO AVE 1    0.000000  0.000000  
1             NaN  309 ANG MO KIO AVE 1    0.000000  0.000000  
2             NaN  309 ANG MO KIO AVE 1    0.000000  0.000000  
3             NaN  309 ANG MO KIO AVE 1    0.000000  0.000000  
4             NaN  216 ANG MO KIO AVE 1  103.841505  1.366197

In [142]:
# Filter out the unique months where remaining lease is NaN
resale_month_NaN = resale_long_lat_fmt_only[resale_long_lat_fmt_only["remaining_lease"].isnull()]["month"]
resale_month_NaN.unique()

array(['1990-01', '1990-02', '1990-03', '1990-04', '1990-05', '1990-06',
       '1990-07', '1990-08', '1990-09', '1990-10', '1990-11', '1990-12',
       '1991-01', '1991-02', '1991-03', '1991-04', '1991-05', '1991-06',
       '1991-07', '1991-08', '1991-09', '1991-10', '1991-11', '1991-12',
       '1992-01', '1992-02', '1992-03', '1992-04', '1992-05', '1992-06',
       '1992-07', '1992-08', '1992-09', '1992-10', '1992-11', '1992-12',
       '1993-01', '1993-02', '1993-03', '1993-04', '1993-05', '1993-06',
       '1993-07', '1993-08', '1993-09', '1993-10', '1993-11', '1993-12',
       '1994-01', '1994-02', '1994-03', '1994-04', '1994-05', '1994-06',
       '1994-07', '1994-08', '1994-09', '1994-10', '1994-11', '1994-12',
       '1995-01', '1995-02', '1995-03', '1995-04', '1995-05', '1995-06',
       '1995-07', '1995-08', '1995-09', '1995-10', '1995-11', '1995-12',
       '1996-01', '1996-02', '1996-03', '1996-04', '1996-05', '1996-06',
       '1996-07', '1996-08', '1996-09', '1996-10', 

Observed that data missing the resale price from 1990 to December 2014.

In [143]:
# Retrieve 5 rows where remaining lease is not null to view how it is calculated.
resale_long_lat_fmt_only[resale_long_lat_fmt_only["remaining_lease"].isnull() == False].head(5)

month        town flat_type block        street_name storey_range  \
656847  2015-01  ANG MO KIO    3 ROOM   174   ANG MO KIO AVE 4     07 TO 09   
656848  2015-01  ANG MO KIO    3 ROOM   541  ANG MO KIO AVE 10     01 TO 03   
656849  2015-01  ANG MO KIO    3 ROOM   163   ANG MO KIO AVE 4     01 TO 03   
656850  2015-01  ANG MO KIO    3 ROOM   446  ANG MO KIO AVE 10     01 TO 03   
656851  2015-01  ANG MO KIO    3 ROOM   557  ANG MO KIO AVE 10     07 TO 09   

        floor_area_sqm      flat_model  lease_commence_date  resale_price  \
656847            60.0        Improved                 1986      255000.0   
656848            68.0  New Generation                 1981      275000.0   
656849            69.0  New Generation                 1980      285000.0   
656850            68.0  New Generation                 1979      290000.0   
656851            68.0  New Generation                 1980      290000.0   

       remaining_lease                address   longitude  latitude  
656847              70   174 ANG MO KIO AVE 4  103.837619  1.375097  
656848              65  541 ANG MO KIO AVE 10  103.855621  1.373922  
656849              64   163 ANG MO KIO AVE 4  103.838176  1.373549  
656850              63  446 ANG MO KIO AVE 10  103.855357  1.367761  
656851              64  557 ANG MO KIO AVE 10  103.857736  1.371626

Observed that the remaining lease requires the current year of row to derive the value.
Next: Column "Year" and "Month" will be added for purposes such as EDA and derivation of values such as "remaining_lease"

## Create Columns for Data Frame

In [144]:
# add month column

In [145]:
# create column with month column formatted to date time format
resale_long_lat_fmt_only['month_fmt'] = pd.to_datetime(resale_long_lat_fmt_only['month'])

In [146]:
# drop "month" coumn
resale_long_lat_fmt_only.drop(['month'], axis=1, inplace = True)

In [147]:
# create month column
resale_long_lat_fmt_only["month"] = resale_long_lat_fmt_only['month_fmt'].dt.month

# create year column
resale_long_lat_fmt_only["year"] = resale_long_lat_fmt_only['month_fmt'].dt.year

In [148]:
# check info and top 3 rows with newly create month and year column

# check info:
print(resale_long_lat_fmt_only.info())

# check first 3 rows:
print(resale_long_lat_fmt_only.head(3))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 903393 entries, 0 to 903392
Data columns (total 16 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   town                 903393 non-null  object        
 1   flat_type            903393 non-null  object        
 2   block                903393 non-null  object        
 3   street_name          903393 non-null  object        
 4   storey_range         903393 non-null  object        
 5   floor_area_sqm       903393 non-null  float64       
 6   flat_model           903393 non-null  object        
 7   lease_commence_date  903393 non-null  int64         
 8   resale_price         903393 non-null  float64       
 9   remaining_lease      194343 non-null  object        
 10  address              903393 non-null  object        
 11  longitude            903393 non-null  float64       
 12  latitude             903393 non-null  float64       
 13  month_fmt     

Observed that year and month column has been created successfully. The original column of month where data is in the format of "yyyy-mm" has been drop.


## Calculate Remaining Lease

Remaining lease is calculated from the difference between 99 and the current age of the house based on current year

In [149]:
resale_long_lat_fmt_only["remaining_lease"] = 99 - (resale_long_lat_fmt_only["year"] 
                                             - resale_long_lat_fmt_only["lease_commence_date"])

In [150]:
# View dataset after calculating remaining lease value:
print(resale_long_lat_fmt_only.info())

# View first 3 rows
print(resale_long_lat_fmt_only.head(3))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 903393 entries, 0 to 903392
Data columns (total 16 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   town                 903393 non-null  object        
 1   flat_type            903393 non-null  object        
 2   block                903393 non-null  object        
 3   street_name          903393 non-null  object        
 4   storey_range         903393 non-null  object        
 5   floor_area_sqm       903393 non-null  float64       
 6   flat_model           903393 non-null  object        
 7   lease_commence_date  903393 non-null  int64         
 8   resale_price         903393 non-null  float64       
 9   remaining_lease      903393 non-null  int64         
 10  address              903393 non-null  object        
 11  longitude            903393 non-null  float64       
 12  latitude             903393 non-null  float64       
 13  month_fmt     

Observed that remaining lease values has been calculated and there is no more null values for  the current values.

# Joined Dataset with Planning areas for cleaning

Planning area is derived from the OneMap website.
The datarows tally with with the original dataset which has been referenced to derive the information

# Import In dataset containing the Planned Areas

In [151]:
# Import in Dataset
planning_area_df = pd.read_csv('../assets/onemap/planning_area_query.csv')


## Check if there are null values in the dataset

In [152]:
# get dataset info
planning_area_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 903393 entries, 0 to 903392
Data columns (total 5 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   month                903393 non-null  object 
 1   town                 903393 non-null  object 
 2   longitude            903393 non-null  float64
 3   latitude             903393 non-null  float64
 4   planning_query_town  804485 non-null  object 
dtypes: float64(2), object(3)
memory usage: 34.5+ MB


Observed that there are missing values for the column "planning_query_town".

This is expected as this column is derived from the onemap api based on given latitude and longitude information.

In [153]:
# Get first 10 rows of data to view relationship between "town" and "planning_query_town"
planning_area_df.head(10)

month        town   longitude  latitude planning_query_town
0  1990-01  ANG MO KIO    0.000000  0.000000                 NaN
1  1990-01  ANG MO KIO    0.000000  0.000000                 NaN
2  1990-01  ANG MO KIO    0.000000  0.000000                 NaN
3  1990-01  ANG MO KIO    0.000000  0.000000                 NaN
4  1990-01  ANG MO KIO  103.841505  1.366197          ANG MO KIO
5  1990-01  ANG MO KIO  103.841667  1.369197          ANG MO KIO
6  1990-01  ANG MO KIO  103.844516  1.368446          ANG MO KIO
7  1990-01  ANG MO KIO  103.836491  1.366824          ANG MO KIO
8  1990-01  ANG MO KIO  103.836491  1.366824          ANG MO KIO
9  1990-01  ANG MO KIO  103.837196  1.368346          ANG MO KIO

## Compare unique values of "town" and "planning_query_town" to observed if there are similarities

In [154]:
# Get unique "town" values
unique_town = planning_area_df["town"].unique()
unique_town

array(['ANG MO KIO', 'BEDOK', 'BISHAN', 'BUKIT BATOK', 'BUKIT MERAH',
       'BUKIT TIMAH', 'CENTRAL AREA', 'CHOA CHU KANG', 'CLEMENTI',
       'GEYLANG', 'HOUGANG', 'JURONG EAST', 'JURONG WEST',
       'KALLANG/WHAMPOA', 'MARINE PARADE', 'QUEENSTOWN', 'SENGKANG',
       'SERANGOON', 'TAMPINES', 'TOA PAYOH', 'WOODLANDS', 'YISHUN',
       'LIM CHU KANG', 'SEMBAWANG', 'BUKIT PANJANG', 'PASIR RIS',
       'PUNGGOL'], dtype=object)

In [155]:
# Get unique "planning_query_town" values
unqiue_town_planning_query = planning_area_df["planning_query_town"].unique()
unqiue_town_planning_query

array([nan, 'ANG MO KIO', 'BEDOK', 'BISHAN', 'BUKIT BATOK', 'BUKIT MERAH',
       'BUKIT TIMAH', 'TANGLIN', 'OUTRAM', 'ROCHOR', 'DOWNTOWN CORE',
       'CHOA CHU KANG', 'CLEMENTI', 'GEYLANG', 'HOUGANG', 'JURONG EAST',
       'JURONG WEST', 'KALLANG', 'CHANGI', 'NOVENA', 'QUEENSTOWN',
       'MARINE PARADE', 'SERANGOON', 'TAMPINES', 'TOA PAYOH', 'WOODLANDS',
       'YISHUN', 'TUAS', 'SUNGEI KADUT', 'LIM CHU KANG', 'BUKIT PANJANG',
       'SEMBAWANG', 'PASIR RIS', 'SENGKANG', 'PUNGGOL'], dtype=object)

In [156]:
# Get unique_town_planning_query without NaN value
len_query= len(unqiue_town_planning_query)
unique_town_planning_query = unqiue_town_planning_query[1:36]

In [157]:
# Get the differences between unqiue town value and unique value of planning_area_town
differences_town = np.setdiff1d(unique_town_planning_query, unique_town)
differences_town

array(['CHANGI', 'DOWNTOWN CORE', 'KALLANG', 'NOVENA', 'OUTRAM', 'ROCHOR',
       'SUNGEI KADUT', 'TANGLIN', 'TUAS'], dtype=object)

- Observed that there are some similarities between the "town" column and the "planning_query_town"
- There are towns found in "planning_query_town" but not "town" such as 'CHANGI', 'DOWNTOWN CORE', 'KALLANG', 'NOVENA', 'OUTRAM', 'ROCHOR','SUNGEI KADUT', 'TANGLIN', 'TUAS'.

In [158]:
# Get town values where "planning_query_town" value is missing
unique_town_missing = planning_area_df[planning_area_df["planning_query_town"].isnull()]["town"].unique()
unique_town_missing


array(['ANG MO KIO', 'BUKIT BATOK', 'BUKIT MERAH', 'CENTRAL AREA',
       'CLEMENTI', 'GEYLANG', 'JURONG EAST', 'JURONG WEST',
       'KALLANG/WHAMPOA', 'QUEENSTOWN', 'SENGKANG', 'TOA PAYOH',
       'WOODLANDS', 'BEDOK', 'LIM CHU KANG', 'YISHUN', 'BISHAN',
       'BUKIT TIMAH', 'CHOA CHU KANG', 'HOUGANG', 'BUKIT PANJANG',
       'MARINE PARADE', 'PASIR RIS', 'SERANGOON', 'TAMPINES', 'SEMBAWANG',
       'PUNGGOL'], dtype=object)

In [159]:
# Get the differences between unqiue town value with missing value and unique value of planning_area_town
# values that are found in derived "planning_query_town" but not in "town"
differences_town_missing = np.setdiff1d(unique_town_planning_query, unique_town_missing)
differences_town_missing

array(['CHANGI', 'DOWNTOWN CORE', 'KALLANG', 'NOVENA', 'OUTRAM', 'ROCHOR',
       'SUNGEI KADUT', 'TANGLIN', 'TUAS'], dtype=object)

In [160]:
# Difference that are found "town" column but not "planning_query_town"
np.setdiff1d(unique_town_missing,unique_town_planning_query)

array(['CENTRAL AREA', 'KALLANG/WHAMPOA'], dtype=object)

In [161]:
# Observerd the unique :town" counts which has the missing "planning_town_query" values
planning_area_df[planning_area_df["planning_query_town"].isnull()]["town"].value_counts()

JURONG WEST        10991
ANG MO KIO          8950
HOUGANG             8093
BEDOK               7557
PASIR RIS           5454
BUKIT MERAH         4679
QUEENSTOWN          4356
YISHUN              4278
GEYLANG             4200
JURONG EAST         4147
KALLANG/WHAMPOA     3775
WOODLANDS           3767
BUKIT BATOK         3646
CLEMENTI            3619
TAMPINES            3526
CHOA CHU KANG       2608
SENGKANG            2312
PUNGGOL             2210
BUKIT PANJANG       2197
TOA PAYOH           2021
BISHAN              1540
SERANGOON           1432
SEMBAWANG           1227
MARINE PARADE       1093
CENTRAL AREA         966
BUKIT TIMAH          232
LIM CHU KANG          32
Name: town, dtype: int64

- Observed that for large number of the town with missing values for planning_query_town, most of the "town" columns map with the "planning_town_query".
- 'CENTRAL AREA', 'KALLANG/WHAMPOA' are columns that are found under "town" column but not in "planning_query_town".

Next:
- KALLANG/WHAMPOA will be mapped to KALLANG,
- Understand data where town is "CENTRAL AREA" to decided what to map data as for "PLANNING_QUERY_TOWN" or to drop the rows.
    - There are 966 rows out of 903393 rows at 0.1%

## Analyse rows where town is "Central Area" and "Planning_Query_Town" is null

In [162]:
# Get rows from planning_area_df where planning_query_town is null
planning_area_df[(planning_area_df["town"] == "CENTRAL AREA") & planning_area_df["planning_query_town"].isnull()].head()

month          town  longitude  latitude planning_query_town
510   1990-01  CENTRAL AREA        0.0       0.0                 NaN
511   1990-01  CENTRAL AREA        0.0       0.0                 NaN
512   1990-01  CENTRAL AREA        0.0       0.0                 NaN
2815  1990-03  CENTRAL AREA        0.0       0.0                 NaN
2816  1990-03  CENTRAL AREA        0.0       0.0                 NaN

Observed that the planning_area_df only shows latitude longitude  
Next: to retrieve unique street name from resale flat price df to better understand the flat locations.

In [163]:
# Get unique street_name from resale flat where town is CENTRAL AREA
resale_long_lat_fmt_only[resale_long_lat_fmt_only["town"] == "CENTRAL AREA"]['street_name'].unique()

array(['JLN KUKOH', 'OUTRAM PK', 'SHORT ST', 'SELEGIE RD', 'UPP CROSS ST',
       'WATERLOO ST', 'QUEEN ST', 'BUFFALO RD', 'ROWELL RD', 'ROCHOR RD',
       'BAIN ST', 'SMITH ST', 'VEERASAMY RD', 'NEW MKT RD',
       'TG PAGAR PLAZA', 'KELANTAN RD', 'SAGO LANE', 'JLN BERSEH',
       'CHANDER RD', 'CHIN SWEE RD', 'KRETA AYER RD', 'OUTRAM HILL',
       'KLANG LANE', 'CANTONMENT RD'], dtype=object)

From the following street name where town is "CENTRAL AREA", the respective town will be mapped: 

- 'JLN KUKOH' will be mapped to 'OUTRAM' as it is near Havelock road.
- 'OUTRAM PK' will be mapped to 'OUTRAM'
- 'SHORT ST' will be mapped to 'ROCHOR' as it is near Rochor station.
- 'SELEGIE RD' will be mapped to 'ROCHOR' as it is near Rochor station.
- 'UPP CROSS ST' will be mapped to 'OUTRAM'
- 'WATERLOO ST' will be mapped to 'ROCHOR'
- 'QUEEN ST' will be mapped to 'ROCHOR'
- 'BUFFALO RD' will be mapped to 'ROCHOR'
- 'ROWELL RD' will be mapped to 'ROCHOR'
- 'ROCHOR RD'will be mapped to 'ROCHOR'
- 'BAIN ST' will be mapped to 'ROCHOR'
- 'SMITH ST' will be mapped to 'OUTRAM'
- 'VEERASAMY RD' will be mapped to 'ROCHOR'
- 'NEW MKT RD' will be mapped to 'OUTRAM'
- 'TG PAGAR PLAZA' will be mapped to 'OUTRAM'
- 'KELANTAN RD' will be mapped to 'ROCHOR'
- 'SAGO LANE' will be mapped to 'OUTRAM'
- 'JLN BERSEH' will be mapped to 'ROCHOR'
- 'CHANDER RD' will be mapped to 'ROCHOR'
- 'CHIN SWEE RD' will be mapped to 'OUTRAM'
- 'KRETA AYER RD'will be mapped to 'OUTRAM'
- 'OUTRAM HILL' will be mapped to 'OUTRAM'
- 'KLANG LANE' will be mapped to 'ROCHOR'
- 'CANTONMENT RD' will be mapped to 'OUTRAM'


In [164]:
# Next: To find out what are the towns with the planning_query_town data missing
planning_area_df[planning_area_df["planning_query_town"].isnull()]


month        town   longitude  latitude planning_query_town
0       1990-01  ANG MO KIO    0.000000  0.000000                 NaN
1       1990-01  ANG MO KIO    0.000000  0.000000                 NaN
2       1990-01  ANG MO KIO    0.000000  0.000000                 NaN
3       1990-01  ANG MO KIO    0.000000  0.000000                 NaN
15      1990-01  ANG MO KIO    0.000000  0.000000                 NaN
...         ...         ...         ...       ...                 ...
903124  2014-12    TAMPINES  103.958598  1.357091                 NaN
903125  2014-12    TAMPINES  103.950536  1.358876                 NaN
903126  2014-12    TAMPINES  103.945141  1.359156                 NaN
903127  2014-12    TAMPINES  103.952874  1.357135                 NaN
903128  2014-12    TAMPINES  103.951613  1.359242                 NaN

[98908 rows x 5 columns]

# Concat Resale Flat Price Data with Town Planning Query

- "planning_town_query" column will be concatenated with with resale flat price dataset.
- The concatenated dataset will have 'street_name' available to be used as a conditional requirement to fill in 'town' column

In [165]:
# Add column "planning_query_town" to datafrane
resale_long_lat_fmt_only_concat_town = resale_long_lat_fmt_only.copy()
resale_long_lat_fmt_only_concat_town["planning_query_town"] = planning_area_df["planning_query_town"]

# view concatenated table
resale_long_lat_fmt_only_concat_town.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 903393 entries, 0 to 903392
Data columns (total 17 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   town                 903393 non-null  object        
 1   flat_type            903393 non-null  object        
 2   block                903393 non-null  object        
 3   street_name          903393 non-null  object        
 4   storey_range         903393 non-null  object        
 5   floor_area_sqm       903393 non-null  float64       
 6   flat_model           903393 non-null  object        
 7   lease_commence_date  903393 non-null  int64         
 8   resale_price         903393 non-null  float64       
 9   remaining_lease      903393 non-null  int64         
 10  address              903393 non-null  object        
 11  longitude            903393 non-null  float64       
 12  latitude             903393 non-null  float64       
 13  month_fmt     

Observed that "planning_town_query" is now added to the dataframe

## Map Planning_Query_Town to Town

In [166]:
# view the rows where the plannning_query_town is NaN
resale_long_lat_fmt_only_concat_town[resale_long_lat_fmt_only_concat_town["planning_query_town"].isnull()].head(5)


town flat_type block       street_name storey_range  floor_area_sqm  \
0   ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12            31.0   
1   ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     04 TO 06            31.0   
2   ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     10 TO 12            31.0   
3   ANG MO KIO    1 ROOM   309  ANG MO KIO AVE 1     07 TO 09            31.0   
15  ANG MO KIO    3 ROOM   247  ANG MO KIO AVE 3     04 TO 06            82.0   

        flat_model  lease_commence_date  resale_price  remaining_lease  \
0         IMPROVED                 1977        9000.0               86   
1         IMPROVED                 1977        6000.0               86   
2         IMPROVED                 1977        8000.0               86   
3         IMPROVED                 1977        6000.0               86   
15  NEW GENERATION                 1978       56900.0               87   

                 address  longitude  latitude  month_fmt  month  year  \
0   309 ANG MO KIO AVE 1        0.0       0.0 1990-01-01      1  1990   
1   309 ANG MO KIO AVE 1        0.0       0.0 1990-01-01      1  1990   
2   309 ANG MO KIO AVE 1        0.0       0.0 1990-01-01      1  1990   
3   309 ANG MO KIO AVE 1        0.0       0.0 1990-01-01      1  1990   
15  247 ANG MO KIO AVE 3        0.0       0.0 1990-01-01      1  1990   

   planning_query_town  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
15                 NaN

### Update rows where town is "KALLANG/WHAMPOA" to "KALLANG " for "planning_query_town".

In [167]:
# Check count where town is "KALLANG/WHAMPOA" and "planning_query_town" is NaN
resale_long_lat_fmt_only_concat_town[
                                    (resale_long_lat_fmt_only_concat_town['town'] == 'KALLANG/WHAMPOA')
                                    & (resale_long_lat_fmt_only_concat_town['planning_query_town'].isnull())
                                    ].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3775 entries, 979 to 902817
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   town                 3775 non-null   object        
 1   flat_type            3775 non-null   object        
 2   block                3775 non-null   object        
 3   street_name          3775 non-null   object        
 4   storey_range         3775 non-null   object        
 5   floor_area_sqm       3775 non-null   float64       
 6   flat_model           3775 non-null   object        
 7   lease_commence_date  3775 non-null   int64         
 8   resale_price         3775 non-null   float64       
 9   remaining_lease      3775 non-null   int64         
 10  address              3775 non-null   object        
 11  longitude            3775 non-null   float64       
 12  latitude             3775 non-null   float64       
 13  month_fmt            3775 non

It is observed that there are 3775 rows to for "planning_query_town" area to be renamed to "KALLANG"

In [168]:
# Rename Column to "KALLANG"
resale_long_lat_fmt_only_concat_town.loc[(resale_long_lat_fmt_only_concat_town['town']=='KALLANG/WHAMPOA')
                                         & (resale_long_lat_fmt_only_concat_town['planning_query_town'].isnull())
                                        ,'planning_query_town'] = 'KALLANG'

# check that column has been updated
resale_long_lat_fmt_only_concat_town[resale_long_lat_fmt_only_concat_town['town']=='KALLANG/WHAMPOA'].head(3)

town flat_type block      street_name storey_range  \
971  KALLANG/WHAMPOA    2 ROOM     9         JLN BATU     07 TO 09   
972  KALLANG/WHAMPOA    3 ROOM     2  ST. GEORGE'S RD     07 TO 09   
973  KALLANG/WHAMPOA    3 ROOM   468    NTH BRIDGE RD     10 TO 12   

     floor_area_sqm flat_model  lease_commence_date  resale_price  \
971            50.0   STANDARD                 1970       15300.0   
972            59.0   IMPROVED                 1976       45700.0   
973            60.0   IMPROVED                 1979       52000.0   

     remaining_lease            address   longitude  latitude  month_fmt  \
971               79         9 JLN BATU  103.787716  1.335736 1990-01-01   
972               85  2 ST. GEORGE'S RD  103.860687  1.323426 1990-01-01   
973               88  468 NTH BRIDGE RD  103.863222  1.305392 1990-01-01   

     month  year planning_query_town  
971      1  1990         BUKIT TIMAH  
972      1  1990             KALLANG  
973      1  1990             KALLANG

In [169]:
# Check if columns have been updated successfully.
# Check count where town is "KALLANG/WHAMPOA" and "planning_query_town" is NaN
resale_long_lat_fmt_only_concat_town[
                                    (resale_long_lat_fmt_only_concat_town['town'] == 'KALLANG/WHAMPOA')
                                    & (resale_long_lat_fmt_only_concat_town['planning_query_town'].isnull())
                                    ].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 0 entries
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   town                 0 non-null      object        
 1   flat_type            0 non-null      object        
 2   block                0 non-null      object        
 3   street_name          0 non-null      object        
 4   storey_range         0 non-null      object        
 5   floor_area_sqm       0 non-null      float64       
 6   flat_model           0 non-null      object        
 7   lease_commence_date  0 non-null      int64         
 8   resale_price         0 non-null      float64       
 9   remaining_lease      0 non-null      int64         
 10  address              0 non-null      object        
 11  longitude            0 non-null      float64       
 12  latitude             0 non-null      float64       
 13  month_fmt            0 non-null      datetime64

Observe and validated that "KALLANG" has been updated for "planning_query_town".

## Update planning_query_town for town "CENTRAL AREA"

- "Central Area" is not an option under "planning_query_town".
- "OUTRAM" or "ROCHOR" will be assigned as a value under "planning_query_town" depending on the "street_name" for town.

In [170]:
# Create List of Street Name where town is "CENTRAL AREA" for planning_query_town to be mapped as "OUTRAM"
street_name_to_OUTRAM = ['JLN KUKOH', 'OUTRAM PK', 'UPP CROSS ST'
                         , 'SMITH ST', 'NEW MKT RD', 'TG PAGAR PLAZA'
                         , 'SAGO LANE', 'CHIN SWEE RD', 'KRETA AYER RD'
                         , 'OUTRAM HILL', 'CANTONMENT RD'
                        ]

# Create List of Street Name where town is "CENTRAL AREA" for planning_query_town to be mapped as "ROCHOR"
street_name_to_ROCHOR = ['SHORT ST', 'SELEGIE RD', 'WATERLOO ST'
                         , 'QUEEN ST', 'BUFFALO RD', 'ROWELL RD'
                         , 'ROCHOR RD', 'BAIN ST', 'VEERASAMY RD'
                         , 'KELANTAN RD', 'JLN BERSEH', 'CHANDER RD'
                         , 'KLANG LANE'                        
                        ] 


In [171]:
# Create for loop to set "planning_query_town" to OUTRAM if street_name matches list

# instantiate counter variable
counter = 0

for street_name in street_name_to_OUTRAM:
    resale_long_lat_fmt_only_concat_town.loc[(resale_long_lat_fmt_only_concat_town['town']=='CENTRAL AREA')
                                             & (resale_long_lat_fmt_only_concat_town['street_name']== street_name)
                                             & (resale_long_lat_fmt_only_concat_town['planning_query_town'].isnull())
                                            ,'planning_query_town'] = 'OUTRAM'

# # Sample check that OUTRAM has been replaced base on street_name
resale_long_lat_fmt_only_concat_town[(resale_long_lat_fmt_only_concat_town['town'] == 'CENTRAL AREA')
                                    & (resale_long_lat_fmt_only_concat_town['street_name'] == 'JLN KUKOH')].head(3)

    

town flat_type block street_name storey_range  floor_area_sqm  \
509   CENTRAL AREA    2 ROOM     8   JLN KUKOH     10 TO 12            53.0   
2827  CENTRAL AREA    3 ROOM     8   JLN KUKOH     13 TO 15            76.0   
3942  CENTRAL AREA    2 ROOM     8   JLN KUKOH     10 TO 12            53.0   

     flat_model  lease_commence_date  resale_price  remaining_lease  \
509    IMPROVED                 1971       30000.0               80   
2827   IMPROVED                 1971       41000.0               80   
3942   IMPROVED                 1971       32500.0               80   

          address   longitude  latitude  month_fmt  month  year  \
509   8 JLN KUKOH  103.839407  1.287378 1990-01-01      1  1990   
2827  8 JLN KUKOH  103.839407  1.287378 1990-03-01      3  1990   
3942  8 JLN KUKOH  103.839407  1.287378 1990-04-01      4  1990   

     planning_query_town  
509               OUTRAM  
2827              OUTRAM  
3942              OUTRAM

In [172]:
# Create for loop to set "planning_query_town" to ROCHOR if street_name matches list

# instantiate counter variable
counter = 0

for street_name in street_name_to_ROCHOR:
    resale_long_lat_fmt_only_concat_town.loc[(resale_long_lat_fmt_only_concat_town['town']=='CENTRAL AREA')
                                             & (resale_long_lat_fmt_only_concat_town['street_name']== street_name)
                                             & (resale_long_lat_fmt_only_concat_town['planning_query_town'].isnull())
                                            ,'planning_query_town'] = 'ROCHOR'

# # Sample check that OUTRAM has been replaced base on street_name
resale_long_lat_fmt_only_concat_town[(resale_long_lat_fmt_only_concat_town['town'] == 'CENTRAL AREA')
                                    & (resale_long_lat_fmt_only_concat_town['street_name'] == 'SHORT ST')].head(3)

town flat_type block street_name storey_range  floor_area_sqm  \
511   CENTRAL AREA    2 ROOM    12    SHORT ST     04 TO 06            38.0   
512   CENTRAL AREA    2 ROOM    13    SHORT ST     04 TO 06            38.0   
2815  CENTRAL AREA    2 ROOM    11    SHORT ST     04 TO 06            38.0   

     flat_model  lease_commence_date  resale_price  remaining_lease  \
511    STANDARD                 1974       22000.0               83   
512    STANDARD                 1974       23200.0               83   
2815   STANDARD                 1974       19000.0               83   

          address  longitude  latitude  month_fmt  month  year  \
511   12 SHORT ST        0.0       0.0 1990-01-01      1  1990   
512   13 SHORT ST        0.0       0.0 1990-01-01      1  1990   
2815  11 SHORT ST        0.0       0.0 1990-03-01      3  1990   

     planning_query_town  
511               ROCHOR  
512               ROCHOR  
2815              ROCHOR

In [173]:
# check that column has been updated
resale_long_lat_fmt_only_concat_town[resale_long_lat_fmt_only_concat_town['town']=='CENTRAL AREA'].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6983 entries, 509 to 902442
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   town                 6983 non-null   object        
 1   flat_type            6983 non-null   object        
 2   block                6983 non-null   object        
 3   street_name          6983 non-null   object        
 4   storey_range         6983 non-null   object        
 5   floor_area_sqm       6983 non-null   float64       
 6   flat_model           6983 non-null   object        
 7   lease_commence_date  6983 non-null   int64         
 8   resale_price         6983 non-null   float64       
 9   remaining_lease      6983 non-null   int64         
 10  address              6983 non-null   object        
 11  longitude            6983 non-null   float64       
 12  latitude             6983 non-null   float64       
 13  month_fmt            6983 non

Observed that there are no more null values for planning_query_town where "town" is 'CENTRAL AREA".

### Update planning_query_town where town is the same  name

In [174]:
# Retrieve unique town name where planning_query_town is null.
resale_long_lat_fmt_only_concat_town[resale_long_lat_fmt_only_concat_town["planning_query_town"].isnull()]["town"].value_counts()


JURONG WEST      10991
ANG MO KIO        8950
HOUGANG           8093
BEDOK             7557
PASIR RIS         5454
BUKIT MERAH       4679
QUEENSTOWN        4356
YISHUN            4278
GEYLANG           4200
JURONG EAST       4147
WOODLANDS         3767
BUKIT BATOK       3646
CLEMENTI          3619
TAMPINES          3526
CHOA CHU KANG     2608
SENGKANG          2312
PUNGGOL           2210
BUKIT PANJANG     2197
TOA PAYOH         2021
BISHAN            1540
SERANGOON         1432
SEMBAWANG         1227
MARINE PARADE     1093
BUKIT TIMAH        232
LIM CHU KANG        32
Name: town, dtype: int64

- Observed that the unique "town" values where "planning_query_town" is NaN has the same name as the existing "planning_query_town" as follows.

- Unique "planning_query_town":
        'ANG MO KIO', 'BEDOK', 'BISHAN', 'BUKIT BATOK', 'BUKIT MERAH',
       'BUKIT TIMAH', 'TANGLIN', 'OUTRAM', 'ROCHOR', 'DOWNTOWN CORE',
       'CHOA CHU KANG', 'CLEMENTI', 'GEYLANG', 'HOUGANG', 'JURONG EAST',
       'JURONG WEST', 'KALLANG', 'CHANGI', 'NOVENA', 'QUEENSTOWN',
       'MARINE PARADE', 'SERANGOON', 'TAMPINES', 'TOA PAYOH', 'WOODLANDS',
       'YISHUN', 'TUAS', 'SUNGEI KADUT', 'LIM CHU KANG', 'BUKIT PANJANG',
       'SEMBAWANG', 'PASIR RIS', 'SENGKANG', 'PUNGGOL'

- Next: update the remaining planning_query_town value with NaN with existing values of "town".

In [175]:
# Instantiate list for town where planning_query_town is NaN.
town_where_planning_query_nan = ["JURONG WEST" ,"ANG MO KIO" , "HOUGANG"
                                , "BEDOK" ,"PASIR RIS" ,"BUKIT MERAH" 
                                ,"QUEENSTOWN" ,"YISHUN" ,"GEYLANG"
                                , "JURONG EAST" ,"WOODLANDS" ,"BUKIT BATOK"
                                , "CLEMENTI" , "TAMPINES" ,"CHOA CHU KANG"
                                , "SENGKANG" ,"PUNGGOL" ,"BUKIT PANJANG"
                                , "TOA PAYOH" , "BISHAN" ,"SERANGOON"
                                , "SEMBAWANG" ,"MARINE PARADE" ,"BUKIT TIMAH"
                                , "LIM CHU KANG"]


# Loop through town list and update NaN values for planning_query_town
for town in town_where_planning_query_nan:
    resale_long_lat_fmt_only_concat_town.loc[(resale_long_lat_fmt_only_concat_town['town']== town)
                                             & (resale_long_lat_fmt_only_concat_town['planning_query_town'].isnull())
                                            ,'planning_query_town'] = resale_long_lat_fmt_only_concat_town['town']

#  Check that there is no more null values for planning_query_town.
resale_long_lat_fmt_only_concat_town.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 903393 entries, 0 to 903392
Data columns (total 17 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   town                 903393 non-null  object        
 1   flat_type            903393 non-null  object        
 2   block                903393 non-null  object        
 3   street_name          903393 non-null  object        
 4   storey_range         903393 non-null  object        
 5   floor_area_sqm       903393 non-null  float64       
 6   flat_model           903393 non-null  object        
 7   lease_commence_date  903393 non-null  int64         
 8   resale_price         903393 non-null  float64       
 9   remaining_lease      903393 non-null  int64         
 10  address              903393 non-null  object        
 11  longitude            903393 non-null  float64       
 12  latitude             903393 non-null  float64       
 13  month_fmt     

In [176]:
# Check count of unique planning_query_town 
resale_long_lat_fmt_only_concat_town["planning_query_town"].value_counts()

TAMPINES         78405
YISHUN           68457
BEDOK            65496
JURONG WEST      65038
WOODLANDS        63789
ANG MO KIO       51158
HOUGANG          49340
BUKIT BATOK      43386
CHOA CHU KANG    37254
BUKIT MERAH      33350
PASIR RIS        32524
TOA PAYOH        30885
SENGKANG         30695
GEYLANG          27647
CLEMENTI         27557
QUEENSTOWN       27531
BUKIT PANJANG    26993
JURONG EAST      24306
SERANGOON        22423
BISHAN           20945
KALLANG          18985
PUNGGOL          18896
SEMBAWANG        12914
MARINE PARADE     7886
NOVENA            5914
BUKIT TIMAH       3674
ROCHOR            3458
OUTRAM            3440
DOWNTOWN CORE      390
CHANGI             248
TANGLIN            171
SUNGEI KADUT       137
LIM CHU KANG        64
TUAS                37
Name: planning_query_town, dtype: int64

- Validated that the planning_town_query column has been NaN value has been filled up.
- The total count of planning_town_query unique column sums up to 903393
- Validated that there are no null values and the data type is accurate.

## Export Cleaned Dataset

In [177]:
# export coordinateList to csv
resale_long_lat_fmt_only_concat_town.to_csv('../assets/data_clean/resale_w_plantown_clean.csv', encoding='utf-8', index=False)

# export as pickle file
resale_long_lat_fmt_only_concat_town.to_pickle('../assets/data_clean/resale_w_plantown_clean.pkl')

- Observed that the files are successfully exported.